src/noise.rs

```
@@ -0,0 +1,452 @@
+//! Noise toolkit.
+//!
+//! Rust bindings follow the original API pretty closely.
+
+use bindings::ffi;
+use bindings::AsNative;
+
+use random::Rng;
+
+/// Available noise types.
+#[repr(C)]
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum NoiseType {
+    Default = ffi::TCOD_NOISE_DEFAULT as isize,
+    Perlin  = ffi::TCOD_NOISE_PERLIN as isize,
+    Simplex = ffi::TCOD_NOISE_SIMPLEX as isize,
+    Wavelet = ffi::TCOD_NOISE_WAVELET as isize,
+}
+
+/// Noise object encapsulates a noise generator.
+#[derive(Debug)]
+pub struct Noise {
+    noise: ffi::TCOD_noise_t,
+    dimensions: u32
+}
+
+/// Default hurst value.
+pub const DEFAULT_HURST: f32 = 0.5;
+
+/// Default lacunarity value.
+pub const DEFAULT_LACUNARITY: f32 = 2.0;
+
+/// Maximum number of octaves for turbulence and fbm noise functions.
+pub const MAX_OCTAVES: u32 = 128;
+
+impl Noise {
+    /// Return an instance of [NoiseInitializer](./struct.NoiseInitializer.html)
+    /// which is used to customize the creation of Noise object.
+    pub fn init_with_dimensions(dimensions: u32) -> NoiseInitializer {
+        NoiseInitializer::new_init_with_dimensions(dimensions)
+    }
+
+    pub fn set_type(&self, noise_type: NoiseType) {
+        unsafe {
+            ffi::TCOD_noise_set_type(self.noise, noise_type as u32)
+        }
+    }
+
+    pub fn get(&self, coords: &mut [f32]) -> f32 {
```

@tomassedovic I don't mind changing `tcod_sys`, and we can prove that the current version included with `tcod-rs` will not mutate `coords`. My problem is that we support other precompiled versions: if any of those (for some reason) change this behaviour things will silently fail. It's not terribly likely to happen, so I'm fine with doing it, but it's something to consider.